### PR TITLE
Mypy check untyped functions

### LINKED ISSUE #5657: Add mypy typchecking of untyped functions for source code (not tests)
**Is your feature request related to a problem? Please describe.**
MyPy currently only checks functions with type definitions. It would be great to add more type annotations to untyped functions, so they are also type checked.

Following config can be added to the `setup.cfg` so that MyPy also checks untyped functions:
```py
check_untyped_defs = True
```
But running MyPy with this setting will show ~1000 type errors.

**Describe the solution you'd like**
Adding type annotations to the codebase so that MyPy runs without errors even when enabling:
```py
check_untyped_defs = True
```

Since most of the type errors happen inside test cases this issue will only include the main source code. Test code can remain dynamically typed without any problems.

**Additional context**
Follow up of #5605.

### PR DESCRIPTION
**Pull request by Shadow-Devil**
Fixes #5657.

### Description

Enable config to also type check untyped functions (not in tests).
- Type annotations were added
- Explicit casts were added
- Variable redefinitions with other types were refactored 
- If no other option was available (or it would be overly verbose) `# type: ignore` was added

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Shadow-Devil** [comment]
I'm a bit confused by the auto fix by pre-commit.ci. I thought that type unions with `A | B` syntax was only added in python 3.10 bit MONAI is targetting versions >= 3.7.
See also here: https://docs.python.org/3/library/typing.html#typing.Union

**Comment by wyli** [comment]
> I'm a bit confused by the auto fix by pre-commit.ci. I thought that type unions with `A | B` syntax was only added in python 3.10 but MONAI is targeting versions >= 3.7. See also here: https://docs.python.org/3/library/typing.html#typing.Union

that particular file has `from __future__ import annotations` which triggers pyupgrade of pre-commit.ci to update the type annotations

**Comment by Shadow-Devil** [comment]
Is it ok to add venv to the exclude path of black or should it be in another PR with an extra issue?
It was just quite annoying since my whole pc froze as black tried to format all dependencies, but I can remove it from this PR.

**Comment by wyli** [comment]
> Is it ok to add venv to the exclude path of black or should it be in another PR with an extra issue? It was just quite annoying since my whole pc froze as black tried to format all dependencies, but I can remove it from this PR.

sure. could you please also avoid `assert` https://deepsource.io/gh/Project-MONAI/MONAI/run/66546629-66ba-4424-8758-2fe9f51ac2ed/python/BAN-B101?listindex=0 you can use `isinstance` to check the types

**Review by wyli — APPROVED** [review]

File: monai/visualize/class_activation_maps.py [hunk 1]
```diff
@@ -204,12 +205,9 @@ def compute_map(self, x, class_idx=None, layer_idx=-1):
 
     def _upsample_and_post_process(self, acti_map, x):
         # upsampling and postprocessing
-        if self.upsampler:
-            img_spatial = x.shape[2:]
-            acti_map = self.upsampler(img_spatial)(acti_map)
-        if self.postprocessing:
-            acti_map = self.postprocessing(acti_map)
-        return acti_map
+        img_spatial = x.shape[2:]
```
Comment (wyli) [inline monai/visualize/class_activation_maps.py:208]: this change will require the previous user's input `upsampler=None` being changed to something like `upsampler=torch.nn.Identity`, but I think it should be a minor change.

**Comment by wyli** [comment]
thanks @Shadow-Devil, I can see that the type annotation enhancements addressed a few potential bugs, that's great.

**Comment by Shadow-Devil** [comment]
@wyli Thank you for apprechiating my work! I would like to go on and as my next type checking enhancement type the deprecated & deprecated_args decorators, since otherwise all functions annotated with them lose type information. Do you think this is a good next step? If yes, I will create an issue and work on it😄

**Comment by wyli** [comment]
thanks, I think those utilities will still need some refactoring (such as for https://github.com/Project-MONAI/MONAI/issues/5653), type checking support could come later after that. But it's a good idea and please help create a feature request...

**Pull request merged**